README.md
# django-ezaddress

Django ezAddress provides an easy and extensible means of handling and 
incorporating address functionality for [Django](https://djangoproject.com) 
sites.

It is heavily inspired by [django-address](https://github.com/furious-luke/django-address).
This is a dulled-down version without the bells and whistles of google-map 
integration and some.


docs/CHANGELOG.md
# Change log

All notable changes to this project will be documented in this file.


## v0.2.0 - [2016-05-19]
### Added
- `AddressMixin`, `GPSMixin` and `AddressGPSMixin` classes to ease 
  defining address fields within models.
- `altitude` and `gps_error` fields to the `Address` model. These fields 
  are equally present within the GSPMixin.

### Removed
- Reverse navigation from states to addresses in the `Address` model.

ezaddress/tests.py
from django.core.exceptions import ValidationError
from django.test import TestCase

from ezaddress.models import Country, State, Address



class BaseTestCase(TestCase):
    
    def setUp(self):
        self.al = Country.objects.create(name='Algeria', code='AL')
        self.zm = Country.objects.create(name='Zimbabwe', code='ZW')
        self.ng = Country.objects.create(name='Nigeria', code='NG')
        self.gh = Country.objects.create(name='Ghana', code='GH')


class CountryTestCase(BaseTestCase):
    
    def test_string_representation(self):
        self.assertEqual('Nigeria', str(self.ng))
    
    def test_countries_ordered_by_name_asc(self):
        items = Country.objects.all()
        self.assertEqual(items.count(), 4)
        self.assertEqual(items[0].code, 'AL')
        self.assertEqual(items[1].code, 'GH')
        self.assertEqual(items[2].code, 'NG')
        self.assertEqual(items[3].code, 'ZW')
    
    def test_validation_fails_for_duplicate_name(self):
        country = Country(name='Nigeria', code='??')
        with self.assertRaises(ValidationError):
            country.full_clean()
    
    def test_validation_fails_for_blank_name(self):
        country = Country(code='??')
        with self.assertRaises(ValidationError):
            country.full_clean()
    
    def test_can_save_country_without_code(self):
        Country.objects.create(name='Senegal')
        self.assertEqual(Country.objects.count(), 5)    # 4 created by setUp


class StateTestCase(BaseTestCase):
    
    def setUp(self):
        super(StateTestCase, self).setUp()
        self.ng_dt = State.objects.create(name='Delta', code='DT', country=self.ng)
        self.ng_ab = State.objects.create(name='Abuja', code='AB', country=self.ng)
        self.ng_lg = State.objects.create(name='Lagos', code='LG', country=self.ng)
        self.gh_ac = State.objects.create(name='Accra', code='AC', country=self.gh)
    
    def test_string_representation(self):
        self.assertEqual('Abuja, Nigeria',  str(self.ng_ab))
    
    def test_string_representation_for_state_without_country(self):
        state = State(name='Abuja')
        with self.assertRaises(Exception):
            str(state)
    
    def test_states_ordered_by_country_state_name_asc(self):
        items = State.objects.all()
        self.assertEqual(items.count(), 4)
        self.assertEqual(items[0].code, 'AC')
        self.assertEqual(items[1].code, 'AB')
        self.assertEqual(items[2].code, 'DT')
        self.assertEqual(items[3].code, 'LG')
    
    def test_validation_fails_for_duplicate_name_for_same_country(self):
        state = State(name='Lagos', code='??', country=self.ng)
        with self.assertRaises(ValidationError):
            state.full_clean()
    
    def test_validation_fails_for_blank_country(self):
        state = State(name='Port Harcourt', code='PH')
        with self.assertRaises(ValidationError):
            state.full_clean()
    
    def test_validation_fails_for_blank_name(self):
        state = State(code='PH', country=self.ng)
        with self.assertRaises(ValidationError):
            state.full_clean()
    
    def test_states_without_code_are_valid(self):
        State.objects.create(name='Port Harcourt', country=self.ng)
        self.assertEqual(State.objects.count(), 5)    # 4 created by setUp


class AddressTestCase(BaseTestCase):
    
    def setUp(self):
        super(AddressTestCase, self).setUp()
        self.ng_dt = State.objects.create(name='Delta', code='DT', country=self.ng)
        self.ng_lg = State.objects.create(name='Lagos', code='LG', country=self.ng)
        self.gh_ac = State.objects.create(name='Accra', code='AC', country=self.gh)
        
        self.addr1 = Address.objects.create(
                # include :: in raw to differentiate from __str__ repr
                raw='No 1 Bank Road, Eko 720015, Lagos, Nigeria ::',
                street='No 1 Bank Road', town_city='Eko', 
                postal_code='720015', state=self.ng_lg)
        self.addr2 = Address.objects.create(
                raw='33 Willams Street, Asaba 720016, Delta, Nigeria',
                street='33 Willams Street', town_city='Asaba',
                postal_code='720016', state=self.ng_lg)
        self.addr3 = Address.objects.create(
                raw='5243 Koffi Avenue, Akanta 982201, Accra, Ghana',
                street='5243 Koffi Avenue', town_city='Akanta',
                postal_code='982201', state=self.gh_ac)
        self.addr4 = Address.objects.create(raw='1 Alu Avenue')
    
    def test_string_representation_with_all_fields_provided(self):
        self.assertEqual('No 1 Bank Road, Eko 720015, Lagos, Nigeria',
                         str(self.addr1))
    
    def test_string_representation_with_only_raw_field(self):
        self.assertEqual('1 Alu Avenue', str(self.addr4))
    
    def test_addresses_ordering(self):
        items = Address.objects.all()
        self.assertEqual(items.count(), 4)
        self.assertEqual(items[0].town_city, '')
        self.assertEqual(items[1].town_city, 'Akanta')
        self.assertEqual(items[2].town_city, 'Asaba')
        self.assertEqual(items[3].town_city, 'Eko')
    
    def test_validation_fails_for_blank_raw_field(self):
        addr = Address(street='No 1 Bank Road', town_city='Eko',
                       postal_code='720015', state=self.ng_lg)
        with self.assertRaises(ValidationError):
            addr.full_clean()
    
    def test_getting_address_as_dict(self):
        fields = ('raw','street','town_city','postal_code')
        addr_dict = self.addr1.as_dict()
        
        for f in fields:
            self.assertEqual(getattr(self.addr1, f), addr_dict.get(f))
        
        state = self.addr1.state
        self.assertEqual(state.name, addr_dict.get('state'))
        self.assertEqual(state.code, addr_dict.get('state_code'))
        
        country = state.country
        self.assertEqual(country.name, addr_dict.get('country'))
        self.assertEqual(country.code, addr_dict.get('country_code'))




ezaddress/tests/__init__.py
"""
ezaddress.tests package
"""

ezaddress/tests/test_forms.py
from django.forms import Form, ValidationError
from django.test import TestCase

from ezaddress.forms import AddressField, AddressWidget
from ezaddress.models import Address



class TestForm(Form):
    address = AddressField()


class AddressFieldTestCase(TestCase):
    
    def setUp(self):
        self.form = TestForm()
        self.field = self.form.base_fields['address']
    
    def test_converting_none_to_address(self):
        self.assertEqual(self.field.to_python(None), None)
    
    def test_converting_empty_string_to_address(self):
        self.assertEqual(self.field.to_python(''), None)
    
    def test_convertion_for_invalid_lat_lng(self):
        self.assertRaises(ValidationError, self.field.to_python, {'latitude': 'x'})
        self.assertRaises(ValidationError, self.field.to_python, {'longitude': 'x'})
    
    def test_convertion_for_empty_lat_lng(self):
        self.assertEqual(self.field.to_python({'latitude': ''}), None)
        self.assertEqual(self.field.to_python({'longitude': ''}), None)
    
    def test_convertion_for_dict_with_only_raw_entry(self):
        addr = self.field.to_python({'raw': 'No 1 Bank Road'})
        self.assertEqual(addr.raw, 'No 1 Bank Road')
    


runtests.py
"""
A stand-alone test runner script, configuring the minimum settings required
for tests to execute.
"""
import os
import sys



# ensure app is found on the import path
BASE_DIR = os.path.join(os.path.dirname(__file__), '..')
sys.path.insert(0, BASE_DIR)

# minimum settings required for app's tests
SETTINGS_DICT = {
    'BASE_DIR': BASE_DIR,
    'INSTALLED_APPS': (
        'ezaddress',
    ),
    'DATABASES': {
        'default': {
            'ENGINE': 'django.db.backends.sqlite3',
            'NAME': os.path.join(BASE_DIR, 'db.sqlite3'),
        },
    },
}


def django_sandbox(func):
    def wrapper():
        # two-step activation to get Django tests to run. First, call 
        # settings.configure() to give Django settings to work with\
        from django.conf import settings
        settings.configure(**SETTINGS_DICT)
        
        # Then, call django.setup() to initialize the application cache
        # and other parts:
        import django
        if hasattr(django, 'setup'):
            django.setup()
        
        func()
    return wrapper


@django_sandbox
def run_tests():
    from django.test.utils import get_runner
    from django.conf import settings
    
    # instantiate a test runner
    TestRunner = get_runner(settings)
    
    # now we run the tests
    test_runner = TestRunner(verbosity=1, interactive=True)
    failures = test_runner.run_tests(['ezaddress.tests'])
    sys.exit(bool(failures))


@django_sandbox
def run_management_command():
    from django.core.management import execute_from_command_line
    execute_from_command_line(sys.argv)



if __name__ == '__main__':
    argv = sys.argv[1:]
    if not argv:
        run_tests()
    else:
        run_management_command()


ezaddress/forms.py
from django.utils.translation import ugettext_lazy as _
from django import forms

from .models import Address, to_address



# python 3 fixes
import sys
if sys.version > '3':
    long = int
    basestring = (str, bytes)
    unicode = str


__all__ = ['AddressWidget', 'AddressField']


class AddressWidget(forms.TextInput):
    pass


class AddressField(forms.ModelChoiceField):
    widget = AddressWidget
    
    def __init__(self, *args, **kwargs):
        kwargs['queryset'] = Address.objects.none()
        super(AddressField, self).__init__(*args, **kwargs)
    
    def to_python(self, value):
        if value is None or value == '':
            return None
        
        # check for garbage lat/lng/alt entries
        for field in ['latitude', 'longitude', 'altitude']:
            if field in value:
                if value[field]:
                    try:
                        value[field] = float(value[field])
                    except:
                        raise forms.ValidationError(
                            _('Invalid value for %(field)s'),
                            code='invalid', params={'field': field})
                else:
                    value[field] = None
        
        # also check for garbage gps_error entry
        field = 'gps_error'
        if field in value:
            try:
                value[field] = int(value[field])
            except:
                raise forms.ValidationError(
                    _('Invalid value for %(field)s'),
                    code='invalid', params={'field': field})
        return to_address(value)



ezaddress/migrations/0001_initial.py
# -*- coding: utf-8 -*-
from __future__ import unicode_literals

from django.db import migrations, models


class Migration(migrations.Migration):

    dependencies = [
    ]

    operations = [
        migrations.CreateModel(
            name='Address',
            fields=[
                ('id', models.AutoField(serialize=False, auto_created=True, primary_key=True, verbose_name='ID')),
                ('street', models.CharField(blank=True, verbose_name='Street Address', max_length=100)),
                ('town_city', models.CharField(blank=True, verbose_name='Town or City', max_length=50)),
                ('postal_code', models.CharField(blank=True, max_length=10)),
                ('raw', models.CharField(max_length=200)),
                ('latitude', models.FloatField(blank=True, null=True)),
                ('longitude', models.FloatField(blank=True, null=True)),
            ],
            options={
                'verbose_name_plural': 'Addresses',
                'ordering': ('state', 'town_city', 'postal_code', 'street'),
            },
        ),
        migrations.CreateModel(
            name='Country',
            fields=[
                ('id', models.AutoField(serialize=False, auto_created=True, primary_key=True, verbose_name='ID')),
                ('name', models.CharField(max_length=50, unique=True)),
                ('code', models.CharField(blank=True, max_length=3)),
            ],
            options={
                'verbose_name_plural': 'Countries',
                'ordering': ('name',),
            },
        ),
        migrations.CreateModel(
            name='State',
            fields=[
                ('id', models.AutoField(serialize=False, auto_created=True, primary_key=True, verbose_name='ID')),
                ('name', models.CharField(max_length=100)),
                ('code', models.CharField(blank=True, max_length=3)),
                ('country', models.ForeignKey(to='ezaddress.Country', related_name='states')),
            ],
            options={
                'ordering': ('country', 'name'),
            },
        ),
        migrations.AddField(
            model_name='address',
            name='state',
            field=models.ForeignKey(related_name='addresses', null=True, blank=True, to='ezaddress.State'),
        ),
        migrations.AlterUniqueTogether(
            name='state',
            unique_together=set([('name', 'country')]),
        ),
    ]


ezaddress/migrations/0002_auto_20160519_1840.py
# -*- coding: utf-8 -*-
from __future__ import unicode_literals

from django.db import migrations, models


class Migration(migrations.Migration):

    dependencies = [
        ('ezaddress', '0001_initial'),
    ]

    operations = [
        migrations.AddField(
            model_name='address',
            name='altitude',
            field=models.FloatField(blank=True, null=True),
        ),
        migrations.AddField(
            model_name='address',
            name='gps_error',
            field=models.PositiveSmallIntegerField(blank=True, null=True),
        ),
        migrations.AlterField(
            model_name='address',
            name='postal_code',
            field=models.CharField(blank=True, max_length=10, verbose_name='Postal Code'),
        ),
        migrations.AlterField(
            model_name='address',
            name='state',
            field=models.ForeignKey(related_name='+', null=True, blank=True, verbose_name='State', to='ezaddress.State'),
        ),
        migrations.AlterField(
            model_name='address',
            name='town_city',
            field=models.CharField(blank=True, max_length=50, verbose_name='Town/City'),
        ),
    ]


ezaddress/models.py
from django.db import models
from django.core.exceptions import ValidationError
from django.utils.translation import ugettext_lazy as _
from django.db.models.fields.related import ForeignObject
try:
    from django.db.models.fields.related_descriptors import ForwardManyToOneDescriptor
except ImportError:
    from django.db.models.fields.related import \
         ReverseSingleRelatedObjectDescriptor as ForwardManyToOneDescriptor
from django.utils.encoding import python_2_unicode_compatible



# python 3 fixes
import sys
if sys.version > '3':
    long = int
    basestring = (str, bytes)
    unicode = str


__all__ = ['Country', 'State', 'Address', 'AddressField']


class InconsistentDictError(Exception):
    pass


def _to_address(value):
    """Converts a dict with proper address keys into an Address object."""
    raw = value.get('raw', '')
    country = value.get('country', '')
    country_code = value.get('country_code', '')
    state = value.get('state', '')
    state_code = value.get('state_code', '')
    postal_code = value.get('postal_code', '')
    street = value.get('street', '')
    town_city = value.get('town_city', '')
    latitude = value.get('latitude', None)
    longitude = value.get('longitude', None)
    altitude = value.get('altitude', None)
    gps_error = value.get('gps_error', None)
    
    # raw value is mandatory; if not present dict isn't an Address equivalent
    if not raw:
        return None
    
    # require both or none of country and state, not either 
    if ((country or state) and not (country and state)):
        raise InconsistentDictError
    
    # handle country
    try:
        country_obj = Country.objects.get(name=country)
    except Country.DoesNotExist:
        if country:
            if len(country_code) > Country._meta.get_field('code').max_length:
                if country_code != country:
                    message_fmt = 'Invalid country code (too long): %s' 
                    raise ValueError(_(message_fmt) % country_code)
                country_code = ''
            country_obj = Country.objects.create(name=country, code=country_code)
        else:
            country_obj = None
    
    # handle state
    try:
        state_obj = State.objects.get(name=state)
    except State.DoesNotExist:
        if state:
            if len(state_code) > State._meta.get_field('code').max_length:
                if state_code != state:
                    message_fmt = 'Invalid state code (too long): %s'
                    raise ValueError(_(message_fmt) % state_code)
                state_code = ''
            state_obj = State.objects.create(name=state, code=state_code, 
                                             country=country_obj)
        else:
            state_obj = None
    
    # handle address
    try:
        if not (street or town_city):
            addr_obj = Address.objects.get(raw=raw)
        else:
            addr_obj = Address.objects.get(
                street = street,
                town_city = town_city,
                postal_code = postal_code,
                state = state_obj,
            )
    except Address.DoesNotExist:
        addr_obj = Address(
            raw = raw,
            street = street,
            town_city = town_city,
            postal_code = postal_code,
            state = state_obj,
            latitude = latitude,
            longitude = longitude,
            altitude = altitude,
            gps_error = gps_error,
        )
    
    addr_obj.save()
    return addr_obj


def to_address(value):
    if value is None:
        return None
    
    if isinstance(value, Address):
        return value
    elif isinstance(value, (int, long)):
        # assume value is model primary key
        return value
    elif isinstance(value, basestring):
        addr_obj = Address(raw = value)
        addr_obj.save()
        return addr_obj
    elif isinstance(value, dict):
        try:
            return _to_address(value)
        except InconsistentDictError:
            return Address.objects.create(raw=value['raw'])
    
    # value not in any of the recognized formats
    raise ValidationError(_('Invalid address value.'))


@python_2_unicode_compatible
class Country(models.Model):
    """A model for storing Country data."""
    name = models.CharField(max_length=50, unique=True, blank=False)
    code = models.CharField(max_length=3, blank=True)
    
    class Meta:
        verbose_name_plural = 'Countries'
        ordering = ('name',)
    
    def __str__(self):
        return self.name


@python_2_unicode_compatible
class State(models.Model):
    """A model for storing State data."""
    name = models.CharField(max_length=100, blank=False)
    code = models.CharField(max_length=3, blank=True)
    country = models.ForeignKey(Country, blank=False, null=False,
                related_name='states')
    
    class Meta:
        unique_together = ('name', 'country')
        ordering = ('country', 'name')
    
    def __str__(self):
        value = self.name
        if self.country and value:
            value += ', '
        value += self.country.name
        return value


@python_2_unicode_compatible
class Address(models.Model):
    raw = models.CharField(max_length=200)
    street = models.CharField(_('Street Address'), max_length=100, blank=True)
    town_city = models.CharField(_('Town/City'), max_length=50, blank=True)
    state = models.ForeignKey(State, verbose_name=_('State'), blank=True, 
                null=True, related_name='+')
    postal_code = models.CharField(_('Postal Code'), max_length=10, blank=True)
    latitude = models.FloatField(blank=True, null=True)
    longitude = models.FloatField(blank=True, null=True)
    altitude = models.FloatField(blank=True, null=True)
    gps_error = models.PositiveSmallIntegerField(blank=True, null=True)
    
    class Meta:
        verbose_name_plural = 'Addresses'
        ordering = ('state', 'town_city', 'postal_code', 'street')
    
    def __str__(self):
        if self.state:
            value = ''
            if self.street:
                value = self.street
            
            if self.town_city:
                if value:
                    value += ', '
                value += self.town_city
                
                if self.postal_code:
                    value += ' %s' % self.postal_code
                
            if value:
                value += ', '
            value += str(self.state)
        else:
            value = self.raw
        return value
    
    def clean(self):
        if not self.raw:
            raise ValidationError(_('Addresses may not have a blank `raw` field.'))
    
    def as_dict(self):
        addr = dict(
            street = self.street,
            town_city = self.town_city,
            raw = self.raw,
            latitude = self.latitude if self.latitude else '',
            longitude = self.longitude if self.longitude else '',
            altitude = self.altitude if self.altitude else '',
            gps_error = self.gps_error if self.gps_error else '',
        )
        if self.state:
            addr['state'] = self.state.name
            addr['state_code'] = self.state.code 
            addr['postal_code'] = self.postal_code
            if self.state.country:
                addr['country'] = self.state.country.name
                addr['country_code'] = self.state.country.code
        return addr


class AddressMixin(object):
    """A mixin which defines address fields for use within a model."""
    addr_street = models.CharField(_('Street Address'), max_length=100, blank=True)
    addr_town   = models.CharField(_('Town/City'), max_length=20, blank=True)
    addr_state  = models.ForeignKey(State, verbose_name=_('State'), blank=True, 
                    null=True, related_name='+')
    postal_code = models.CharField(_('Postal Code'), max_length=10, blank=True)


class GPSMixin(object):
    """A mixin which defines GPS lat/lng/alt fields for use within a model."""
    latitude = models.FloatField(blank=True, null=True)
    longitude = models.FloatField(blank=True, null=True)
    altitude  = models.FloatField(blank=True, null=True)
    gps_error = models.PositiveSmallIntegerField(blank=True, null=True)


class AddressGPSMixin(AddressMixin, GPSMixin):
    """A mixin which defines address and geometric lat/lng/alt fields for use 
    within a model.
    """
    pass


class AddressDescriptor(ForwardManyToOneDescriptor):
    
    def __set__(self, instance, value):
        super(AddressDescriptor, self).__set__(instance, to_address(value))


class AddressField(models.ForeignKey):
    description = 'An address'
    
    def __init__(self, **kwargs):
        kwargs['to'] = 'ezaddress.Address'
        super(AddressField, self).__init__(**kwargs)
    
    def contribute_to_class(self, cls, name, virtual_only=False):
        super(ForeignObject, self).contribute_to_class(cls, name, virtual_only)
        setattr(cls, self.name, AddressDescriptor(self))

